### ISSUE TITLE
In cart quantity modified on shopping cart page update

### ISSUE THREAD

**Issue by mc12345678** [zencart/zencart#1163]
In an update performed in commit https://github.com/zencart/zencart/commit/7f975f39e154a75fac5e63f9cfcca9040ceb8e84
as a part of [pull 92](https://github.com/zencart/zencart/pull/92) an attempted fix for notifying customers that the quantity in the cart had been modified was applied.  The code works fine when product is added to the cart, but when the cart is updated (pressing the "refresh" button while in the cart). If the quantity of product that was in the cart plus the quantity that has been entered (even if unchanged) is greater than the existing stock of that product, then the quantity would be modified (reduced). If there were two product in the cart when this update was performed a sort of "chase" condition would occur where one is "corrected" the other would go wrong, try to fix, and role would reverse.

This is discussed on the ZC site at [this thread](https://www.zen-cart.com/showthread.php?220786-Update-Button-of-Shopping-Cart-Reduces-Quantity-in-Cart).

**Comment by netcrop** [comment]
I agree the issue still remain. The following test set up has been made on v160:
* StoreBackend:

Product has a quantity of 10 available.
Product is set to have a maximum of 9.
Product is not mixed quantity.
Configuration->Stock:Allow checkout = false

* StoreFront:

Shopping cart: add Qty 10
Refresh: Qty became 9, that is correct.
Warning: The quantity added to your cart has been adjusted because of a restriction on maximum you are allowed.

Shopping cart: add Qty 9
Refresh: Qty became 1, that is correct. But the warning shouldn't shows up.
Warning: Quantity has been adjusted to what is in stock.

Shopping cart: add Qty 6
Refresh: Qty became 4, that is correct.But the warning shouldn't shows up.
Warning: Quantity has been adjusted to what is in stock.

Shopping cart: add Qty 7
Refresh: Qty became 6, that is wrong.
Warning: Quantity has been adjusted to what is in stock.

Shopping cart: add Qty 8
Refresh: Qty became 3, that is wrong.
Warning: Quantity has been adjusted to what is in stock.

Conclusion: Unexpected Behaviour. Need further investigation.

**Comment by mc12345678** [comment]
Point of note based on the above test by @netcrop. 
With regards to this portion of the test:
> Shopping cart: add Qty 9
Refresh: Qty became 1, that is correct. But the warning shouldn't shows up.
Warning: Quantity has been adjusted to what is in stock.

I don't see how the quantity becoming 1 could be correct after the refresh. I say this from the perspective that the data shown and/or entered while on the shopping cart page represents the total quantity applicable to the purchase not an addition to or removal of a numbered quantity. 

By the value becoming 1, it is as if a quantity of 8 has been removed from the total future purchase. 

The shopping cart page in a default install does not/has not operated like the product page where changing the number associated with a product and refreshing changes the total quantity by that number. The code specifically prevents a quantity less than 0, therefore could not enter -1 to reduce the quantity from 9 to 8. Also the value of 0 upon cart refresh works to remove the product from the cart entirely. Instead the expectation is that the quantity desired to finally purchase is the number that should exist to the side of the product when the shopping cart page is visited.

**Comment by netcrop** [comment]
* This is what actionUpdateProduct() function carry out:

X1 is a initial qty in cart, which is giving as 1.
X2 is customer filled new qty in her cart.
Total is product qty in total, which is 10 in this test case.
```
if X2 <= Total - X1
        Refresh: X2
else if X2 > Total - X1
        Refresh: Total - X1

12 > 10 - 1
Refresh: 9
11 > 9 # from previous refresh qty.
Refresh: 1
13 > 1 # from previous refresh qty.
Refresh: 9
5 < 9 # from previous refresh qty.
Refresh: 5
24 > 5 # from previous refresh qty.
Refresh: 5
```
* What we have here isn't Unexpected qty, but from previous refreshed qty. 
* Some code inside function  actionUpdateProduct() prevent customer to fill in a max qty that should be allowed.

**Comment by mc12345678** [comment]
The code that has been removed as shown in the commit is the code that introduced the problem being seen. It was properly considered when the quantity "entered" represents by how much to change the total such as when adding product to the cart (on the product info page, one is identifying the quantity to which should be added to the cart.) But in an update, the quantity provided is how much should *be* in the cart not a difference. 

In review of the initial referenced code changes it was incorporated with good intentions and if the update *did* identify the quantity by which it should change it would be correct. But an honest mistake was made in the process. 

As to what the code is doing, surprisingly the update function is actually relatively small. Collect data, test for various conditions based on store and product setup, take applicable action for the quantities and the significant part of the introduced change, notify the customer about what was happening... 

This particular condition was likely overlooked if using a "busy" test site. By busy meaning having a large quantity of product available and making a relatively "normal" test purchase. I haven't gone to look at the quantities normally available in the test store setup and even if there are a few product with low quantities, the likelihood of testing with that one or two product is low. When using a product with a large quantity (even 100) it would be unlikely that the test would include updating the cart with a quantity of 50+ if the original consideration were that one would purchase possibly no more than 5 or 10. Therefore this response would not have been seen. 

I happened to be trying to test basically a rewritten version of the actionUpdateProduct code to support another plugin. When I was getting an unusual response, I went back to the base code. I think I might have seen this issue a year ago, but I wasn't as determined to ensure ZC operated correctly and I thought for sure that I was wrong in the expected operation or that I was "alone" in identifying it or that if/when it got fixed that I would have to incorporate the change. My perspective is different now. :)

**Comment by mc12345678** [comment]
The code that has been removed as shown in the commit is the code that introduced the problem being seen. It was properly considered when the quantity "entered" represents by how much to change the total such as when adding product to the cart (on the product info page, one is identifying the quantity to which should be added to the cart.) But in an update, the quantity provided is how much should *be* in the cart not a difference. 

In review of the initial referenced code changes it was incorporated with good intentions and if the update *did* identify the quantity by which it should change it would be correct. But an honest mistake was made in the process. 

As to what the code is doing, surprisingly the update function is actually relatively small. Collect data, test for various conditions based on store and product setup, take applicable action for the quantities and the significant part of the introduced change, notify the customer about what was happening... 

This particular condition was likely overlooked if using a "busy" test site. By busy meaning having a large quantity of product available and making a relatively "normal" test purchase. I haven't gone to look at the quantities normally available in the test store setup and even if there are a few product with low quantities, the likelihood of testing with that one or two product is low. When using a product with a large quantity (even 100) it would be unlikely that the test would include updating the cart with a quantity of 50+ if the original consideration were that one would purchase possibly no more than 5 or 10. Therefore this response would not have been seen. 

I happened to be trying to test basically a rewritten version of the actionUpdateProduct code to support another plugin. When I was getting an unusual response, I went back to the base code. I think I might have seen this issue a year ago, but I wasn't as determined to ensure ZC operated correctly and I thought for sure that I was wrong in the expected operation or that I was "alone" in identifying it or that if/when it got fixed that I would have to incorporate the change. My perspective is different now. :)

**Comment by netcrop** [comment]
I agree more test cases needs to be made with multiple products in cart before you close this issue. I'll come back with comments.

**Comment by netcrop** [comment]
* I suggest a refactoring of function actionUpdateProduct() and add_cart() because these two functions doing is_numeric() validition, which could be done implicit by:
```
Replace
<input type="text" name="">.
with HTML5
<input type="number" name="" min="min qty" max="max qty">.
```
* involved Files:
```
includes/classes/shopping_cart.php
includes/functions/html_output.php
includes/modules/pages/shopping_cart/header_php.php
```

**Comment by mc12345678** [comment]
While the above comment appears to have been edited a couple of times, to address the current status of having a zen_draw_input_number_field, the text field is not hardcoded into the function zen_draw_input_field, it is instead the default to be presented.  
```php
function zen_draw_input_field($name, $value = '', $parameters = '', $type = 'text', $reinsert_value = true) {
    $field = '<input type="' . zen_output_string($type) . '" name="' . zen_sanitize_string(zen_output_string($name)) . '"';
```

And therefore to present a tag such as 

`<input type="number" name="" min="min qty" max="max qty">`
one need to echo the function such as:
```php
echo zen_draw_input_field('myname', $last_value, (isset($min_qty) ? ' min="' . $min_qty . '"' : '') . (isset($max_qty) ? ' max="'.$max_qty . '"' : ''), 'number', true); 

I'll go with the thought though, so if there were a function zen_draw_input_number_field which as stated appears to be HTML5 compliant, what would that function do if the template were not HTML5 and how would it know? 

The next thing with the min="min qty" and max="max qty", besides that this would output a small amount more text per page load of an info page or other locations where quantity could be entered by the customer, on each such page load the actual remaining max quantities would need to be calculated and potentially recalculated while options were chosen on the page.  

As far as the benefits, the fact that the function (whether using zen_draw_input_field with a $type = "number" or using a new function zen_draw_input_number_field) would output such type = "number" and potentially min or max, someone visiting the webpage could modify that information using an easily accessible browser tool before submitting/POSTing the data to be processed.  Therefore, the check of is_numeric() would still need to be performed to prevent breaking the cart or otherwise performing a malicious action.  As said, the above suggestions do make for a better customer experience and could/should be incorporated into one's template, but for example ZC 1.5.5 was provided with an HTML5 template, it does not "require" one to be used and if it were not so flexibly written to always ensure the operation of the "base" install, then a non HTML5 template would break or allow breaking the cart if the is_numeric check inside those functions were removed.

I would also say that in regards to the min/max being pushed to the browser field(s), again an added bonus but for example the application in which this issue was discovered provides for potential different min and max quantities based on other selections made on the page.  If for example on one selection the max quantity were 8, the customer had entered 6, and then the selection were revised to one that had a maximum of 4, how would/should that be addressed? What if they return to the previous selection?  

The way I see it, the value would have to be stored, the maximum updated, the user entered value modified to the new max (refresh of field or whole page), then if the selection were reversed back to the original, should the customer be presented again with the same "4" or restored back to what was originally entered? Would think that if quantities are so automatically modified, the user should also be notified of the change so that they are aware that they will not be getting what they originally had entered.  This is also assuming that the visitor has allowed some form of javascript to be active, otherwise the values would be stuck and with that in mind such code that modifies the min/max would have to assume that such modification could not be made and set the min/max to the outer extents of such setting to allow for selection of any/all actually available quantities.

**Comment by netcrop** [comment]
![updateproduct](https://cloud.githubusercontent.com/assets/20638400/17292671/222518c4-581e-11e6-86d7-e9c27a1e4268.png)
![add_cart](https://cloud.githubusercontent.com/assets/20638400/17292685/2a516ca0-581e-11e6-98f1-22ab4139ea0c.png)
These are the results look like, I have tried in my setup. it works.

**Comment by mc12345678** [comment]
> Refer to your very first post for this issue:


> > If there were two product in the cart when this update was performed a sort of "chase" condition would occur where one is "corrected" the other would go wrong, try to fix, and role would reverse.

> Do you think my formula 2 and 3 from previous post gives a answer to this issue?

Umm.. It is difficult to say that it would or wouldn't.  The reason that the "race" issue I describe would occur is because of the additional check that exists in the current code of if the $new_qty + $cart_qty > $chk_current_qty to reset the $new_qty to a lower value that would not violate that (incorrect) relationship.  In an update situation, the new_qty is expected to be the final value to be applied to the cart regardless of how many items were previously in the cart.  Therefore at least in one condition addressed by this area of the code, if $new_qty > $chk_current_qty would be a reason to modify the total quantity.  A majority of the other reasons are already identified in the switch (true) section.  I've identified a few other conditions that are not directly addressed at this point of the processing (non-mixed product that has attributes, can add the total quantity of the product to the cart for each attribute combination, but it doesn't seem that until checkout_confirmation is accessed does it become an issue.  But having looked through the code a bit in relation to such a "setup", identification of the "problem" product is difficult because of the sequential review.  Sure the action could occur sooner, but there are also a couple of things that could be done to prevent such an issue: mixed attributes for one, as well as using maximum quantities as another for particular product.  

So to fully see what I am talking about regarding the "race" condition. As before, two separate products, each has a maximum value that is less than the total stock.  Add the first product to the cart in an amount that if added to itself would exceed the maximum.  Goto the cart. Do not modify anything. Select the update button.  Now, go to the second product.  Again add a quantity that if added to itself would exceed the maximum. 
Goto the cart.
Take note of the quantities present.  In the example of using 9 as the maximum, if originally 9 were added, then refreshed the current qty becomes 1, so in this current view of the cart, would see 1 of the first product and 9 of the second product.  Attempt to update the 1 to a 9.  Refresh the page, now the first product shows a quantity of 8 (not 9) and the second product shows 1.  modify the second product again to try to get 9, then refresh and they swap again.  To break this loop, one must/could enter 1 for the product that shows 8, then refresh and both product's 1 and 2 will show a quantity of 1.  Now enter 9 and then refresh and both will switch to 8 (never reaching 9 again while doing an update).  But if the code is revised as reflected in #1164, then the above issue is completely removed, the messages still are displayed about how the cart was modified, but no "abnormal" modification is made.

Does it require application of equations 2 and/or 3? Perhaps they could apply, though I think there is more logic and conditional statements involved in that than the current arrangement.  Yes the current code isn't exactly straight forward from a what should be changed perspective, but it is compact and one can see the effects of it if moving down the path of the code with a single object/product.  I have a sort of summary/analysis of the code that I've been working on, but am not complete with so not yet posting here. But will offer what I see/have seen.  If you see a way to make an improvement, you are always welcome to work it out on your own fork of the code and make your proposal for the ZC team to evaluate.  In fact that is the reason I posted what I did, because it seems like it's not right, but then I found what I thought was a solution to the local problem.  I have another item of improvement that could/would help to prevent a store from being setup to experience this, but that's a separate issue. :)

**Comment by netcrop** [comment]
* In function actionUpdateProduct(), this variable has been defined few times but no where it has been used?
```
 $adjust_new_qty = 'true'; 
```

**Comment by mc12345678** [comment]
To that, in the proposed solution, that variable is removed for two reasons. 1) no action is performed that relates to the considered reason for it to be present (no alteration of the price, just the adjustment of as related to the max value), 2) there has been no use of the variable in the code other than potential local use as the variable is not made global nor presented in a notifier.

That said, the same variable appears in actionAddProduct and actionMultipleAddProduct. In both functions the variable is set in the same logic, but again it is not shared with anything else. Result is in order to take advantage of it, a modification to the function is required either by overriding it or directly modifying it. In either case a core file is affected. But also, this is unrelated to the observed and reported problem.

**Comment by netcrop** [comment]
> 1) no action is performed that relates to the considered reason for it to be present (no alteration of the price, just the adjustment of as related to the max value)
* The price in browser do change.

**Comment by mc12345678** [comment]
> These two replacement functions have been used to test this issue. I think it may give some help.

> ```php
function update_quantity($products_id, $quantity) {
    global $messageStack;
    $this->contents[] = array($products_id);
        $refresh = $_SESSION['refresh'];
        if($refresh <= 0){
            $refresh = zen_get_products_quantity_order_max($products_id);
        }
        $_SESSION['refresh'] = $quantity;
        $current = $quantity;
        $stock = zen_get_products_stock($products_id);
        $max = zen_get_products_quantity_order_max($products_id);
        if ($current <= $stock - $refresh){
            if ($current <= $max){
                $refresh = $current;
            } else {
                $refresh = $max;
            }
        } else {
            if ($refresh >= $stock - $max){
                if($stock - $refresh > 0){
                    $refresh = $stock - $refresh;
                }else{
                    $refresh = $max;
                }
            }else{
                $refresh = $max;
            }
        }
        $messageStack->add_session('header', ($this->display_debug_messages ? 'FUNCTION ' . __FUNCTION__ . ': ' : '') . "$current,$refresh", 'success');
        $this->contents[$products_id] = array('qty' => (float)$refresh);
    }
```
> ```php
function actionUpdateProduct($goto, $parameters) {
     for ($i=0, $n=sizeof($_POST['products_id']); $i<$n; $i++) {
         $new_qty = $_POST['cart_quantity'][$i];
         $this->update_quantity($_POST['products_id'][$i], $new_qty);
     }
}
```

Curious, how does this complete rewrite of two functions which only addresses product that are not mixed help with "testing. This issue"? What is the overall problem that is being attempted to resolve where the above code would affect every plugin that uses these two functions and would have to take this complete rewrite into account? 

The problem identified in this documented issue is resolved by [this change](https://github.com/zencart/zencart/pull/1164/files).

Overall, whether pieced together or originally so designed, the various action functions are written to address the differences encountered by taking the applicable action and then they feed the result to a central process of add_cart. The values received by add_cart are the final state quantities. From there, additional verification is made that is/can be consistent with all actions to be taken in actionxxx code.

**Comment by mc12345678** [comment]
> > 1) no action is performed that relates to the considered reason for it to be present (no alteration of the price, just the adjustment of as related to the max value)
> ```php 
$this->contents[$products_id] = array('qty' => (float)$refresh);
```
> - The useful result comes out by update $refresh variable in browser.
> > 2) there has been no use of the variable in the code other than potential local use as the variable is not made global nor presented in a notifier.
> - No notifier has been triggered because the purpose is to varifying logic. Same reason no updating dabase has been involved.
> - Instead of weave in all in one 100 line function. I start with to dedicate different responsibillty to different functions. more coming.

To this:
> - No notifier has been triggered because the purpose is to varifying logic. Same reason no updating dabase has been involved.

I disagree. The two are unrelated. A notification would allow other code to know that such a change occurred and be able to take applicable/desired action rather than the contents to change within a blackbox but with no output provided until all changes have been made as established by the base code. Sure for others reading this it is possible to override any part or the entire class, but that may not be necessary as much with additional data/access made available.

Modifying the database is merely capturing the data/condition that has been identified as desired to keep or maintain and is also a separate operation/handler and in this discussion is cmpletely unrelated to why/how the quantity in the cart was modified when pressing the update button in ZC versions 1.5.2 through 1.5.5.

**Comment by netcrop** [comment]
> Curious, how does this complete rewrite of two functions which only addresses product that are not mixed help with "testing. This issue"?

* Testing was based on your original post: [Product is not mixed quantity] (https://www.zen-cart.com/showthread.php?220786-Update-Button-of-Shopping-Cart-Reduces-Quantity-in-Cart).
* I agree my replacement test functions are far from complete. Due to the complexity of the original code, I choose to rewirte the two functions for testing purpose and for understanding the code.
* You may close this issue as you like. I think the complexity of the call chain actionUpdateProduct - add_cart - update_quantity - ajust_quantity with mixed SQL statements slow down development process and cause maintaince headache.

**Comment by mc12345678** [comment]
I've tried countless times to provide additional response to this thread; however, in almost every case some form of interruption has occurred that ultimately has caused the loss of what I have written.

I've reread through this (yet again) and if able would like to provide a summary of what I expect and what I perceive has otherwise been discussed.

**Knowns/Constants:**
This issue can be easily reproduced using realistic quantities and settings. ie. A product has only a certain quantity, the store does not allow "backorder", a maximum has been applied to the product:
The store has the following settings in the configuration->Stock section:
> Check stock level 	true 	 
Subtract stock 	true 	
Allow Checkout 	false 	

Which is a minor change to a default out-of-the-box install of ZC with sample product.

Product has to be limited by a maximum quantity of any amount between half of the available quantity of the product up to something less than the total quantity available of the product and not be 0.

In a default ZC store with sample product, there is no default product that could be used directly to test this condition.  At least one characteristic of the product to be tested must be changed.  Whether it be the quantity of available product or the maximum quantity to be set or modified.  (My suggestion to see the full affect of the situation is to edit either products_id 1 or 2 to provide a maximum that is between 16 and 30.  Let' say to make the change to products_id 2 (Matrox G400 32MB) for consistency. To see that the message is unnecessarily presented or that values evaluated don't make sense then set the maximum to something equal to or greater than the quantity of existing product.)  

Default store assigns a quantity of 31 for each of those products_id, Problem can be easily seen with a maximum set between 1/2 of the total quantity (rounded up to nearest number of decimals set for product quantity in admin->configuration->stock) and less than the total quantity of 31 entered (ie. between 16 and 30), but the values remain unchanged (message still displayed that the quantity was adjusted) if the maximum is set higher than or equal to the available quantity of 31 or higher.
After setup of the store and until purchase is made, quantities/maximums and other settings of the product remain unchanged by the store owner.  There are known/expected situations to occur, but really want to keep this down to the bare essentials to reproduce and correct this issue which remains in ZC 1.5.5b.

**Expectation:**
After product has been added to the cart and before paying for the cart contents, I believe it should be possible to review the contents and where necessary update the quantity to be purchased directly in the shopping cart.  I believe that the quantity to be entered on the shopping cart page should represent the total quantity desired of that unique product and that the entry made is not to identify by how much to increase or decrease the total quantity to be purchased.  Therefore if I originally added a quantity of say 2 of an item, then as I was looking at the shopping cart and decided I wanted to actually receive a quantity of 5, I should enter 5 as the quantity and select refresh/update or similar action button and the total should now show 5 at least in the very simplest of conditions.  This expectation is also furthered by the text that popsup when hovering over the button: 

> Change your quantity by highlighting the number in the box, correcting the quantity and clicking this button.

So if using the setup described above in the **Knowns/Constants:** section, then product 2 (Matrox G400 32MB) has a maximum that is set somewhere between 16 and 30, the stock is 31, you have an empty cart.  

Now to see how the store/cart "should" work, select product 1 (Matrox G200 MMS).  For the situation to be seen by all in the test, add a quantity of 16 to the cart. It doesn't matter what attributes as long as an attribute is selected.
Go to the shopping cart if not automatically taken there. Select the update button (refresh circle on the shopping cart screen, or any ZC provided refresh/update option), notice that the page comes back with no warnings and that the quantity of the product (Matrox G200 MMS) in the cart is still 16.  Do that refresh as much as desired.  Even feel free to increase the number of the product up to or even over the actual maximum available (31) just to see how the cart responds.  This is what I would call normal/expected behavior.

Now: the "trouble".  goto products_id 2 (Matrox G400 32MB), now add 16 of the product to the cart.
Again refresh/update the cart... Your quantity of Matrox G400 32MB is not 16 anymore... But your quantity of Matrox G200 MMS is still whatever you left it at.  Go ahead, change the quantity to match the maximum allowed for the product, refresh/update the cart, Still doesn't get updated to that value... In fact, there is no way to update the cart and get the maximum value, instead the product must be directly added from the store and then update never performed again from the shopping cart page after that.

The problem comes from the code in the actionUpdateProduct section:

```php
        if ($add_max != 0) {
// bof: adjust new quantity to be same as current in stock
            if (STOCK_ALLOW_CHECKOUT == 'false' && ($new_qty + $cart_qty > $chk_current_qty)) {
                $adjust_new_qty = 'true';
                $alter_qty = $chk_current_qty - $cart_qty;
                $new_qty = ($alter_qty > 0 ? $alter_qty : 0);
                $messageStack->add_session('shopping_cart', ($this->display_debug_messages ? 'FUNCTION ' . __FUNCTION__ . ': ' : '') . WARNING_PRODUCT_QUANTITY_ADJUSTED . zen_get_products_name($_POST['products_id'][$i]), 'caution');
            }
// eof: adjust new quantity to be same as current in stock
```

Where:
```php
        $add_max = zen_get_products_quantity_order_max($_POST['products_id'][$i]); // maximum allowed
        $cart_qty = $this->in_cart_mixed($_POST['products_id'][$i]); // total currently in cart
        $new_qty = $_POST['cart_quantity'][$i]; // new quantity
        $current_qty = $this->get_quantity($_POST['products_id'][$i]); // how many currently in cart for attribute
          // bof: ... quantity ... current in stock // mc12345678 Comment of this line edited (...) to provide clarification for current discussion.
          $chk_current_qty = zen_get_products_stock($_POST['products_id'][$i]);
```
Such that $new_qty is the value a specific product has at the point of selecting the update/refresh button. So it works out that if there is a maximum number of purchasable product in the cart and then the cart is updated, if the number of product existing in the cart plus the number of product chosen for the update quantity exceeds the available quantity of the product then the customer is at least incorrectly notified that the quantity has been changed and if the maximum allowed is less than the total available, then the quantity can be incorrectly modified when that same summation is experienced.  When a maximum is applied, the "new_qty" on an update should only be modified if the quantity entered exceeds the total available (this is already addressed in code above) or if it exceeds the maximum allowable (incorrectly addressed for the update action).  A solution to this is as previously stated provided at pr #1164.  

As to the previous post:
> > Curious, how does this complete rewrite of two functions which only addresses product that are not mixed help with "testing. This issue"?
> - Testing was based on your original post: [Product is not mixed quantity](https://www.zen-cart.com/showthread.php?220786-Update-Button-of-Shopping-Cart-Reduces-Quantity-in-Cart).
> - I agree my replacement test functions are far from complete. Due to the complexity of the original code, I choose to rewirte the two functions for testing purpose and for understanding the code.
> - You may close this issue as you like. I think the complexity of the call chain actionUpdateProduct - add_cart - update_quantity - ajust_quantity with mixed SQL statements slow down development process and cause maintaince headache.

I see that you have generated some new version of the action related code as a separate issue #1198, and commend you for going through development to possibly make the operation easier to read at least on the surface, my thing is I don't see any difficulty with "development" or maintenance or perhaps don't see how having the code rewritten would have with 100% certainty prevented adding the small amount of code that causes the above issue.  

I only expect to see this issue or a similar closed when the observed problem has been corrected.  I thought perhaps the problem would be confirmed and potential solution validated, but instead an entirely different way of handling the shopping cart was proposed to address an error that was accidentally introduced a while ago.  It is unfortunate that it did not get resolved in the latest release of ZC, but perhaps further improvement will come in whatever shape.
